M2/Macaulay2/packages/MinimalPrimes.m2

```diff
@@ -196,36 +195,25 @@ minimalPrimes Ideal := List => opts -> I -> minprimesHelper(I, (minimalPrimes, I
 decompose     Ideal := List => options minimalPrimes >> opts -> I -> minprimesHelper(I, (minimalPrimes, Ideal), opts)
 
 -- keys: none so far
-MinimalPrimesOptions = new SelfInitializingType of BasicList
-MinimalPrimesOptions.synonym = "minimal primes options"
+MinimalPrimesContext = new SelfInitializingType of Context
+MinimalPrimesContext.synonym = "minimal primes context"
 
 -- keys: CodimensionLimit and Result
-MinimalPrimesComputation = new Type of MutableHashTable
+MinimalPrimesComputation = new Type of Computation
 MinimalPrimesComputation.synonym = "minimal primes computation"
 
--- if there is a compatible computation stored in I.cache,
--- returns the computation object, otherwise creates the entry:
---   MinimalPrimesOptions{} => MinimalPrimesComputation{ CodimensionLimit, Result }
-new MinimalPrimesComputation from Ideal  := (C, I) -> (
-    -- TODO: currently there are no options that could go in MinimalPrimesOptions,
-    -- but can we give options to this function? This is needed for saturate, etc.
-    cacheKey := MinimalPrimesOptions{};
-    try I.cache#cacheKey else I.cache#cacheKey = new MinimalPrimesComputation from {
-	Result => null, CodimensionLimit => -1 })
-
-isComputationDone = method(TypicalValue => Boolean, Options => true)
+new MinimalPrimesComputation from Ideal := (C, I) -> new MinimalPrimesComputation from {
+    CodimensionLimit => -1,
+    Result           => null}
+
 isComputationDone MinimalPrimesComputation := Boolean => options minimalPrimes >> opts -> container -> (
     -- this function determines whether we can use the cached result, or further computation is necessary
-    try instance(container.Result, List) and opts.CodimensionLimit <= container.CodimensionLimit else false)
+    instance(container.Result, List)
+    and opts.CodimensionLimit <= container.CodimensionLimit)
 
-cacheComputation = method(TypicalValue => CacheFunction, Options => true)
-cacheComputation MinimalPrimesComputation := CacheFunction => options minimalPrimes >> opts -> container -> new CacheFunction from (
-    -- this function takes advantage of FunctionClosures by modifying the container
-    computation -> (
-	if isComputationDone(opts, container) then ( cacheHit class container; container.Result ) else
-	if (result := computation(opts, container)) =!= null then (
-	    container.CodimensionLimit = opts.CodimensionLimit;
-	    container.Result = result)))
+updateComputation(MinimalPrimesComputation, List) := List => options minimalPrimes >> opts -> (container, result) -> (
```

Comment:
A note: we would like our computations to be interruptable, and then be able to restart them.  It seems we cannot do that yet for minimal primes with this mechanism (Context, Computation being added here).  Is that correct?

The actual code for computing minimal primes is too recursive currently, making it hard to interrup and restart, or even view what was found up to that point  (I really need to make this a git issue)

Reply:
I think this is possible. See `code (associatedPrimes, Module)`, where `L` is the mutable list containing the partial results. Two things are necessary:
- `isComputationDone` should tell when the computation needs to keep going
- the algorithms should use the cached results and store their partial results

Reply:
Actually I'm not certain if the `associatedPrimes` algorithm uses cached partial results, but it definitely stores partial results. @jchen419 do you know?

Reply:
I think `associatedPrimes` would use cached partial results in at least 2 places: i) for partial resolution (if `CodimensionLimit` is given), and ii) for minimal primes (again with `CodimensionLimit`)

Reply:
You're right, and now that I look at the algorithm for modules again, it caches after every codimension is complete but not in the middle of a given codimension, which makes sense.

Reply:
@mahrud We can handle this (or think about it) later, after the pull request, is that ok with you?